File: Daily/MARCH/20_03_3108/3108.java
import java.util.*;

class Solution {
    public int[] minimumCost(int n, int[][] edges, int[][] queries) {
        Map<Integer, List<int[]>> graph = new HashMap<>();
        for (int i = 0; i < n; i++) {
            graph.put(i, new ArrayList<>());
        }
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int w = edge[2];
            graph.get(u).add(new int[]{v, w});
            graph.get(v).add(new int[]{u, w});
        }

        int[] component = new int[n];
        Arrays.fill(component, -1);
        Map<Integer, Integer> componentMinAnd = new HashMap<>();
        int compId = 0;

        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();

        for (int startNode = 0; startNode < n; startNode++) {
            if (component[startNode] == -1) {
                queue.offer(startNode);
                component[startNode] = compId;
                int minAnd = -1; // Initialize with -1 (all bits set)
                visited.clear();
                visited.add(startNode);

                while (!queue.isEmpty()) {
                    int node = queue.poll();
                    for (int[] neighborInfo : graph.get(node)) {
                        int neighbor = neighborInfo[0];
                        int weight = neighborInfo[1];
                        if (!visited.contains(neighbor)) {
                            visited.add(neighbor);
                            component[neighbor] = compId;
                            queue.offer(neighbor);
                        }
                        minAnd &= weight;
                    }
                }
                componentMinAnd.put(compId, minAnd);
                compId++;
            }
        }

        int[] result = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int s = queries[i][0];
            int t = queries[i][1];
            if (component[s] != component[t]) {
                result[i] = -1;
            } else {
                result[i] = componentMinAnd.get(component[s]);
            }
        }

        return result;
    }
}

File: Daily/MARCH/20_03_3108/3108.py
from collections import defaultdict, deque

class Solution:
    def minimumCost(self, n, edges, queries):

        graph = defaultdict(list)
        for u, v, w in edges:
            graph[u].append((v, w))
            graph[v].append((u, w))


        component = [-1] * n 
        component_min_and = {} 
        comp_id = 0

        def bfs(start):
            queue = deque([start])
            component[start] = comp_id
            min_and = -1 

            visited = set([start])

            while queue:
                node = queue.popleft()
                for neighbor, weight in graph[node]:
                    if neighbor not in visited:
                        visited.add(neighbor)
                        component[neighbor] = comp_id
                        queue.append(neighbor)
                    min_and &= weight  

            component_min_and[comp_id] = min_and


        for node in range(n):
            if component[node] == -1:  
                bfs(node)
                comp_id += 1

        result = []
        for s, t in queries:
            if component[s] != component[t]: 
                result.append(-1)
            else:
                result.append(component_min_and[component[s]])

        return result
